ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByCombiner.java

```diff
@@ -0,0 +1,397 @@
+  // "file." to avoid the config being filtered out.
+  private ReduceWork getReduceWork(TaskContext context) {
+    String destVertexName;
+    if (context instanceof TezOutputContextImpl) {
+      destVertexName = ((TezOutputContextImpl)context).getDestinationVertexName();
+    } else {
+      // As of now only map side combiner is supported.
+      return null;
+    }
+
+    String plan =  conf.get(getConfigPrefix(destVertexName) +
+            HiveConf.ConfVars.PLAN.varname);
+    if (plan == null) {
+      LOG.info("Reduce plan is not set for vertex " + destVertexName);
+      return null;
+    }
+    this.conf.set(HiveConf.ConfVars.PLAN.varname, plan);
+    if (conf.getBoolean(getConfigPrefix(destVertexName)
+                    + HiveConf.ConfVars.HIVE_RPC_QUERY_PLAN.varname,
+            true)) {
+      Path planPath = new Path(plan);
+      planPath = new Path(planPath, REDUCE_PLAN_NAME);
+      String planString = conf.get(getConfigPrefix(destVertexName) +
+              planPath.toUri().getPath());
+      this.conf.set(planPath.toUri().getPath(), planString);
+      this.conf.set(HiveConf.ConfVars.HIVE_RPC_QUERY_PLAN.varname, "true");
+    } else {
+      this.conf.set(HiveConf.ConfVars.HIVE_RPC_QUERY_PLAN.varname, "false");
+    }
+    this.conf.set(HAS_REDUCE_WORK, "true");
+    this.conf.set(MAPRED_REDUCER_CLASS, ExecReducer.class.getName());
+
+    return Utilities.getReduceWork(conf);
+  }
+
+  private void initObjectInspectors(TableDesc valueTableDesc,TableDesc keyTableDesc)
+          throws SerDeException {
+    inputKeyDeserializer =
+            ReflectionUtils.newInstance(keyTableDesc.getSerDeClass(), null);
+    inputKeyDeserializer.initialize(null, keyTableDesc.getProperties(), null);
+    keyObjectInspector = inputKeyDeserializer.getObjectInspector();
+
+    keyStructInspector = (StructObjectInspector) keyObjectInspector;
+    firstValueColumnOffset = keyStructInspector.getAllStructFieldRefs().size();
+
+    AbstractSerDe inputValueDeserializer = (AbstractSerDe) ReflectionUtils.newInstance(
+            valueTableDesc.getSerDeClass(), null);
+    inputValueDeserializer.initialize(null, valueTableDesc.getProperties(), null);
+    valueObjectInspector = inputValueDeserializer.getObjectInspector();
+    valueStructInspectors = (StructObjectInspector) valueObjectInspector;
+    numValueCol = valueStructInspectors.getAllStructFieldRefs().size();
+  }
+
+  void initVectorBatches() throws HiveException {
+    inputBatch = batchContext.createVectorizedRowBatch();
+
+    // Create data buffers for value bytes column vectors.
+    for (int i = firstValueColumnOffset; i < inputBatch.numCols; i++) {
+      ColumnVector colVector = inputBatch.cols[i];
+      if (colVector instanceof BytesColumnVector) {
+        BytesColumnVector bytesColumnVector = (BytesColumnVector) colVector;
+        bytesColumnVector.initBuffer();
+      }
+    }
+
+    batchValueDeserializer =
```

Comment:
split this: to deserializeRead and vectorDeserializeRow
also: VectorizedBatchUtil.typeInfosFromStructObjectInspector(valueStructInspectors) could be reused below